File: lua/lush/windows.lua
package.preload["lush.windows"] = function()
  local windows = {}

  function windows.open(name)
    _lush_window_open(name)
  end

  function windows.close(name)
    _lush_window_close(name)
  end

  function windows.toggle(name)
    _lush_window_toggle(name)
  end

  function windows.show(name)
    _lush_window_set_visible(name, true)
  end

  function windows.hide(name)
    _lush_window_set_visible(name, false)
  end

  function windows.set_visible(name, visible)
    _lush_window_set_visible(name, visible and true or false)
  end

  function windows.is_visible(name)
    return _lush_window_is_visible(name)
  end

  function windows.list()
    return _lush_window_list()
  end

  return windows
end

File: src/config/runtime_api/windows.rs
use mlua::{Lua, Result};

use crate::runtime::lua_runtime::AppCommand;
use crate::runtime::lua_runtime::LuaStateBridge;

pub fn install(lua: &Lua, bridge: LuaStateBridge) -> Result<()> {
    let open_bridge = bridge.clone();
    lua.globals().set(
        "_lush_window_open",
        lua.create_function(move |_, name: String| {
            open_bridge.queue_app_command(AppCommand::Open(name));
            Ok(())
        })?,
    )?;

    let close_bridge = bridge.clone();
    lua.globals().set(
        "_lush_window_close",
        lua.create_function(move |_, name: String| {
            close_bridge.queue_app_command(AppCommand::Close(name));
            Ok(())
        })?,
    )?;

    let toggle_bridge = bridge.clone();
    lua.globals().set(
        "_lush_window_toggle",
        lua.create_function(move |_, name: String| {
            toggle_bridge.queue_app_command(AppCommand::Toggle(name));
            Ok(())
        })?,
    )?;

    let visible_bridge = bridge.clone();
    lua.globals().set(
        "_lush_window_set_visible",
        lua.create_function(move |_, (name, visible): (String, bool)| {
            visible_bridge.queue_app_command(AppCommand::SetVisible(name, visible));
            Ok(())
        })?,
    )?;

    let is_visible_bridge = bridge.clone();
    lua.globals().set(
        "_lush_window_is_visible",
        lua.create_function(move |_, name: String| Ok(is_visible_bridge.is_window_visible(&name)))?,
    )?;

    lua.globals().set(
        "_lush_window_list",
        lua.create_function(move |lua, (): ()| {
            let out = lua.create_table()?;
            for (idx, name) in bridge.window_names().into_iter().enumerate() {
                out.set(idx + 1, name)?;
            }
            Ok(out)
        })?,
    )?;

    Ok(())
}
